File: Problems/No.1876_Substrings_of_Size_Three_with_Distinct_Characters/substrings_of_size_three_with_distinct_characters.py
class Solution(object):
    def countGoodSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """

        # parameters
        count, duplicated = 0, 0
        check_count = {}
        
        # start computing
        if (len(s) < 3):
            return 0
        else:
            for i in range(len(s)):
                check_count[s[i]] = check_count.get(s[i], 0) + 1
                if (check_count[s[i]] == 2):
                    duplicated += 1
                
                if (i < 2):
                    continue
                else:
                    if not duplicated:
                        count += 1
                    check_count[s[i-2]] -= 1
                    if (check_count[s[i-2]] == 1):
                        duplicated -= 1

            return count

File: Problems/No.1876_Substrings_of_Size_Three_with_Distinct_Characters/substrings_of_size_three_with_distinct_characters.cpp
class Solution {
public:
    int countGoodSubstrings(string s) {
        
        // parameters
        int count = 0;
        int duplicated = 0;
        std::unordered_map<char, int> check_count;

        // start computing
        if (s.length() < 3) return 0;
        else {

            for (int i=0; i<s.length(); i++) {

                check_count[s[i]]++;
                if (check_count[s[i]] == 2) duplicated++;

                if (i < 2) continue;
                else {

                    if (duplicated == 0) count++;

                    check_count[s[i-2]]--;
                    if (check_count[s[i-2]] == 1) duplicated--;

                }

            }

            return count;

        }

    }
};
